### PR TITLE
 Hods 176 : Added number of employees endpoint and tests surrounding it. 

### PR DESCRIPTION
**Pull request by parkeradam**

### CODE REVIEW

File: conf/app_version_v1_0.routes [hunk 1]
```diff
@@ -1 +1,2 @@
-GET        /corporation-tax                        @uk.gov.hmrc.organisationsdetailsapi.controllers.CorporationTaxController.corporationTax(matchId: java.util.UUID)
\ No newline at end of file
+GET        /corporation-tax                        @uk.gov.hmrc.organisationsdetailsapi.controllers.CorporationTaxController.corporationTax(matchId: java.util.UUID)
+POST        /number-of-employees                    @uk.gov.hmrc.organisationsdetailsapi.controllers.NumberOfEmployeesController.numberOfEmployees(matchId: java.util.UUID)
```
Comment (Luka-Abey) [inline conf/app_version_v1_0.routes:2]:
```suggestion
POST       /number-of-employees                    @uk.gov.hmrc.organisationsdetailsapi.controllers.NumberOfEmployeesController.numberOfEmployees(matchId: java.util.UUID)
```

**Review by Luka-Abey — CHANGES_REQUESTED** [review]
just very minor style things - looks good :)

**Review by Luka-Abey — APPROVED** [review]

File: app/uk/gov/hmrc/organisationsdetailsapi/controllers/NumberOfEmployeesController.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  def numberOfEmployees(matchId: UUID): Action[JsValue] = Action.async(bodyParsers.json) {
+    implicit request =>
+      authenticate(scopesService.getEndPointScopes("number-of-employees"), matchId.toString) { authScopes =>
+        withValidJson[EmployeeCountRequest] { employeeCountRequest =>
+          val correlationId = validateCorrelationId(request)
+
+          numberOfEmployeesService.get(matchId, employeeCountRequest, authScopes).map { numberOfEmployees =>
+            val selfLink = HalLink("self", s"$self?matchId=$matchId")
+
+            val response = Json.toJson(state(Json.obj("returns" -> numberOfEmployees)) ++ selfLink)
```
Comment (ajchilts) [inline app/uk/gov/hmrc/organisationsdetailsapi/controllers/NumberOfEmployeesController.scala:56]: Should this be `returns`? Documentation shows counts. As this is already in the Json do we want this here?

File: app/uk/gov/hmrc/organisationsdetailsapi/services/NumberOfEmployeesService.scala [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  def get(matchId: UUID, employeeCountRequest: EmployeeCountRequest, scopes: Iterable[String])
+         (implicit hc: HeaderCarrier, request: RequestHeader, ec: ExecutionContext): Future[Option[Seq[NumberOfEmployeesResponse]]] = {
+    resolve(matchId).flatMap {
+      organisationMatch =>
+        val fieldsQuery = scopesHelper.getQueryStringFor(scopes.toList, "number-of-employees")
+        val cacheKey = scopesService.getValidFieldsForCacheKey(scopes.toList)
+        cacheService
+          .get(
+            cacheId = CorporationTaxCacheId(matchId, cacheKey),
```
Comment (ajchilts) [inline app/uk/gov/hmrc/organisationsdetailsapi/services/NumberOfEmployeesService.scala:47]: CorporationTaxCacheId? I think we need a new cache ID here as it needs to also include the date ranges past in. I.e. a request may be considered different with different date ranges.

**Review by ajchilts — CHANGES_REQUESTED** [review]

**Review by ajchilts — APPROVED** [review]
LGTM. Well done 👍

**Review by Luka-Abey — APPROVED** [review]

**Pull request merged**